src/float32-scaling.jl

```diff
@@ -68,11 +70,80 @@ function f32_convert_matrix(ls::LinearScaling, space::Symbol)
 end
 inv_f32_convert_matrix(ls::LinearScaling, space::Symbol) = f32_convert_matrix(inv(ls), space)
 
+is_identity_transform(ls::LinearScaling) = (ls.scale == Vec3d(1)) && (ls.offset == Vec3d(0))
+is_identity_transform(ls::Nothing) = true # Float32Convert with scaling == nothing is neutral/identity
 
-# returns Matrix R such that M * ls = ls * R
-patch_model(::Nothing, M::Mat4d) = Mat4f(M)
-function patch_model(ls::LinearScaling, M::Mat4d)
-    return Mat4f(f32_convert_matrix(ls) * M * f32_convert_matrix(inv(ls)))
+"""
+    patch_model(plot)
+    patch_model(plot, f32c, model)
+
+The (default) order of operations is: 
+
+1. `plot.transformation.transform_func`
+2. `plot.transformation.model`
+3. `scene.float32convert`
+4. `camera.projectionview`
+
+But we want to apply the `float32convert` before `model` so that that can be 
+applied on the GPU. This function evaluates if this is possible and returns an 
+adjusted `LinearScaling` Observable for `apply_transform_and_f32_conversion()`
+and an adjusted `model` matrix Observable for the GPU.
+"""
+patch_model(@nospecialize(plot)) = patch_model(plot, f32_conversion(plot), plot.model)
+
+function patch_model(@nospecialize(plot), f32c::Nothing, model::Observable)
+    return Observable(nothing), map(Mat4f, model)
+end
+
+# TODO: How do we actually judge this well?
+function is_float_safe(scale, trans)
+    resolution = 1e4
+    return all(scale .> resolution .* eps.(Float32.(trans)))
+end
+
+function patch_model(@nospecialize(plot), f32c::Float32Convert, model::Observable) # Observable{Any} :(
+    f32c_obs  = Observable{LinearScaling}(f32c.scaling[], ignore_equal_values = true)
+    model_obs = Observable{Mat4f}(Mat4f(I), ignore_equal_values = true)
+
+    onany(plot, f32c.scaling, model, update = true) do f32c, model
+        # Neutral f32c can mean that data and model cancel each other and we 
+        # still have Float32 preicsion issues inbetween.
+
+        # works with rotation component as well
+        trans, scale = decompose_translation_scale_matrix(model)
+        is_rot_free = is_translation_scale_matrix(model)
+
+        if is_float_safe(scale, trans) && is_identity_transform(f32c)
+            # model should not have Float32 Problems and f32c can be skipped
+            # (model can have rotation here)
+            f32c_obs[] = f32c
+            model_obs[] = Mat4f(model)
```

Comment:
Pulling model translation and scaling into the float32convert (and probably also just applying model on CPU) messes with render order if plots are at equal z. This should fix that and maybe also improve performance a bit